applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp

```diff
@@ -0,0 +1,367 @@
+        r_val_1[0] = values[counter++];
+        r_val_1[1] = values[counter++];
+        r_val_1[2] = values[counter++];
+        r_val_2[0] = values[counter++];
+        r_val_2[1] = values[counter++];
+        r_val_2[2] = values[counter++];
+    }
+}
+
+void recvDataField_doubleVector_DefaultName(ModelPart& rModelPart, const Variable< array_1d<double, 3> >& rVariable1, const Variable< array_1d<double, 3> >& rVariable2)
+{
+    const std::string name = rVariable1.Name() + rVariable2.Name();
+    recvDataField_doubleVector(rModelPart, name, rVariable1, rVariable2);
+}
+
+void sendMesh(const ModelPart& rModelPart, const std::string rName, const bool UseConditions)
+{
+    // extract information from ModelPart
+    const int numNodes = rModelPart.NumberOfNodes();
+    const int numElems = (UseConditions) ? (rModelPart.NumberOfConditions()) : (rModelPart.NumberOfElements());
+
+    std::vector<double> nodes(numNodes*3);
+    std::vector<int> nodeIDs(numNodes);
+    std::size_t node_counter = 0;
+    for (const auto& r_node : rModelPart.Nodes()) {
+        const auto& r_coords = r_node.GetInitialPosition(); // TODO or current coords?
+        nodes[node_counter*3]   = r_coords[0];
+        nodes[node_counter*3+1] = r_coords[1];
+        nodes[node_counter*3+2] = r_coords[2];
+        nodeIDs[node_counter++] = r_node.Id();
+    }
+
+    std::vector<int> numNodesPerElem(numElems);
+    std::vector<int> elems;
+    elems.reserve(numElems*3);
+    std::size_t elem_counter = 0;
+    if (UseConditions) {
+        for (const auto& r_elem : rModelPart.Conditions()) {
+            const auto& r_geom = r_elem.GetGeometry();
+            numNodesPerElem[elem_counter++] = r_geom.PointsNumber();
+            for (const auto& r_node : r_geom) {
+                elems.push_back(r_node.Id());
+            }
+        }
+    } else {
+        for (const auto& r_elem : rModelPart.Elements()) {
+            const auto& r_geom = r_elem.GetGeometry();
+            numNodesPerElem[elem_counter++] = r_geom.PointsNumber();
+            for (const auto& r_node : r_geom) {
+                elems.push_back(r_node.Id());
+            }
+        }
+    }
+
+    EMPIRE_API_sendMesh(const_cast<char*>(rName.c_str()), numNodes, numElems, &nodes[0], &nodeIDs[0], &numNodesPerElem[0], &elems[0]);
+}
+
+void sendMesh_DefaultName(ModelPart& rModelPart, const bool UseConditions)
+{
+    sendMesh(rModelPart, const_cast<char*>(rModelPart.Name().c_str()), UseConditions);
+}
+
+void recvMesh(ModelPart& rModelPart, const std::string rName, const bool UseConditions)
+{
+    KRATOS_ERROR_IF(rModelPart.NumberOfNodes() > 0) << "ModelPart is not empty, it has nodes!" << std::endl;
+    KRATOS_ERROR_IF(rModelPart.NumberOfProperties() > 0) << "ModelPart is not empty, it has properties!" << std::endl;
+    KRATOS_ERROR_IF(rModelPart.IsDistributed()) << "ModelPart cannot be distributed!" << std::endl;
+
+    int numNodes;
+    int numElems;
+    double** nodes = new double*;
```

Comment:
This should be like this 
``` cpp
int** a = new int*[rowCount];
for(int i = 0; i < rowCount; ++i)
    a[i] = new int[colCount];
```

Reply:
https://stackoverflow.com/questions/936687/how-do-i-declare-a-2d-array-in-c-using-new